File: src/main/java/phoenix/world/structures/RemainsStructure.java
package phoenix.world.structures;

import mcp.MethodsReturnNonnullByDefault;
import net.minecraft.util.Rotation;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.MutableBoundingBox;
import net.minecraft.world.biome.Biome;
import net.minecraft.world.gen.ChunkGenerator;
import net.minecraft.world.gen.feature.NoFeatureConfig;
import net.minecraft.world.gen.feature.structure.ScatteredStructure;
import net.minecraft.world.gen.feature.structure.Structure;
import net.minecraft.world.gen.feature.structure.StructureStart;
import net.minecraft.world.gen.feature.template.TemplateManager;

import javax.annotation.ParametersAreNonnullByDefault;

@MethodsReturnNonnullByDefault
@ParametersAreNonnullByDefault
public class RemainsStructure extends ScatteredStructure<NoFeatureConfig>
{
    public RemainsStructure()
    {
        super(NoFeatureConfig::deserialize);
    }

    @Override
    public String getStructureName()
    {
        return "Remains";
    }

    @Override
    public int getSize()
    {
        return 3;
    }

    @Override
    public Structure.IStartFactory getStartFactory()
    {
        return Start::new;
    }

    @Override
    protected int getSeedModifier()
    {
        return 14357618;
    }

    public static class Start extends StructureStart
    {
        public Start(Structure<?> structure, int chunkPosX, int chunkPosZ, MutableBoundingBox mbb, int references, long seed)
        {
            super(structure, chunkPosX, chunkPosZ, mbb, references, seed);
        }

        public void init(   ChunkGenerator<?> generator,    TemplateManager templateManagerIn, int chunkX, int chunkZ,    Biome biomeIn)
        {
            int i = chunkX * 16;
            int j = chunkZ * 16;
            BlockPos blockpos = new BlockPos(i, 90, j);
            Rotation rotation = Rotation.values()[this.rand.nextInt(Rotation.values().length)];
            ErasedPieces.init(templateManagerIn, blockpos, rotation, this.components, this.rand);
            this.recalculateStructureSize();
        }
    }
}

File: src/main/java/phoenix/world/structures/remains/RemainsStructure.kt
package phoenix.world.structures.remains

import com.mojang.datafixers.Dynamic
import net.minecraft.util.Rotation
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.MutableBoundingBox
import net.minecraft.world.biome.Biome
import net.minecraft.world.gen.ChunkGenerator
import net.minecraft.world.gen.feature.NoFeatureConfig
import net.minecraft.world.gen.feature.structure.ScatteredStructure
import net.minecraft.world.gen.feature.structure.Structure
import net.minecraft.world.gen.feature.structure.Structure.IStartFactory
import net.minecraft.world.gen.feature.structure.StructureStart
import net.minecraft.world.gen.feature.template.TemplateManager
import phoenix.world.structures.remains.RemainsPieces.init
import java.util.function.Function

class RemainsStructure :
    ScatteredStructure<NoFeatureConfig>(Function { p_214639_0_: Dynamic<*> ->
        NoFeatureConfig.deserialize(
            p_214639_0_
        )
    })
{
    override fun getStructureName(): String = "Remains"

    override fun getSize(): Int = 3

    override fun getStartFactory(): IStartFactory
    {
        return IStartFactory{
                structure: Structure<*>, chunkPosX: Int, chunkPosZ: Int, mbb: MutableBoundingBox, references: Int, seed: Long ->
            Start(structure, chunkPosX, chunkPosZ, mbb, references, seed)
        }
    }

    override fun getSeedModifier(): Int = 14357618

    class Start(structure: Structure<*>, chunkPosX: Int, chunkPosZ: Int, mbb: MutableBoundingBox, references: Int, seed: Long) : StructureStart(structure, chunkPosX, chunkPosZ, mbb, references, seed)
    {
        override fun init(
            generator: ChunkGenerator<*>,
            templateManagerIn: TemplateManager,
            chunkX: Int,
            chunkZ: Int,
            biomeIn: Biome
        )
        {
            val i = chunkX * 16
            val j = chunkZ * 16
            val blockpos = BlockPos(i, 90, j)
            val rotation = Rotation.values()[rand.nextInt(Rotation.values().size)]
            init(generator, templateManagerIn, blockpos, rotation, components, rand)
            recalculateStructureSize()
        }
    }
}